### PR TITLE
refactor: Clean up client/request package

### LINKED ISSUE #2425: Clean up client/request package

### PR DESCRIPTION
**Pull request by AndrewSisley**
## Relevant issue(s)

Resolves #2425

## Description

Cleans up client/request package, removing some extra stuff, and documenting most of the remaining.

It is not perfect, but it should hopefully be better.

### CODE REVIEW

File: client/request/select.go [hunk 1]
```diff
@@ -12,39 +12,36 @@ package request
...
-	GroupBy immutable.Option[GroupBy]
-	Filter  immutable.Option[Filter]
-
+// ChildSelect represents a type with selectable child properties.
+//
+// At least one child must be selected.
+type ChildSelect struct {
+	// Fields contains the set of child properties to return.
+	//
+	// At least one child propertt must be selected.
```
Comment (fredcarle) [inline client/request/select.go:43]: todo: `property`

Comment (AndrewSisley) [reply client/request/select.go:43]: Lol

- [ ] Fix typo

File: client/request/select.go [hunk 2]
```diff
@@ -111,25 +108,20 @@ func (s *Select) validateGroupBy() []error {
...
-
-	// Properties above this line match the `Select` object and
-	// are deserialized using the normal/default logic.
-	// Properties below this line require custom logic in `UnmarshalJSON`
-	// in order to be deserialized correctly.
-
-	Fields []map[string]json.RawMessage
 }
 
 func (s *Select) UnmarshalJSON(bytes []byte) error {
```
Comment (fredcarle) [inline client/request/select.go:127]: thought: Since `ChildSelect` has a custom `Unmarshaller`, this should now be unecessary.

**Review by fredcarle — APPROVED** [review]
Thank you for adding all that documentation and even creating the `Fooable`s. It adds to the clarity of what the request types contain.

File: client/request/select.go (hunk 2 above)
Comment (AndrewSisley) [reply client/request/select.go:127]: Was documented in the commit message:

> The use of a ChildSelect object allows the custom code for desrialising fields to live on a dedicated object. Sadly, due to a very annoying 'quirk' of the json lib, embedded UnmarshalJSON functions are not picked up, and so we still have to do some manual property mapping of the parent.

Comment (fredcarle) [reply client/request/select.go:127]: It does pick it up. I added print statements in the ChildSelect unmarshal method and it does get in there. Something else is behaving weirdly though.

Comment (fredcarle) [reply client/request/select.go:127]: So it's not that the emvbedded UnmarshalJSON is not picked up. It's that it ignores the other fields: https://stackoverflow.com/a/38982567

Comment (fredcarle) [reply client/request/select.go:127]: Either way, you have to keep it like it is now. All good 👍.

**Pull request merged**